### PR TITLE
 ng_sixlowpan: make lookup in reassembly buffer more efficient

### PR DESCRIPTION
**Pull request by Osblouf**
Hi, after looking into the code, I think this should lessen a little the computing charge.
With this, the code computes fastest checks first (tag, size) and, if they pass, it computes most expensive tests (memcmp calls). It also checks first the tag as it can be a good solution to discriminate easily two different packets and then, other tests confirms that it is the right packet. 
This is really fast compared to bloom filter as it add computations when adding and removing an entry into the buffer. With the defined buffer size (4U), it does not worth it.

### CODE REVIEW

**Comment by cgundogan** [comment]
The only difference is that you check tag before size now, or am I mistaken?

File: sys/net/gnrc/network_layer/sixlowpan/frag/rbuf.c [hunk 1]
```diff
@@ -268,27 +268,29 @@ static rbuf_t *_rbuf_get(const void *src, size_t src_len,
...
+                    (unsigned)rbuf[i].pkt->size, rbuf[i].tag);
+                rbuf[i].arrival = now.seconds;
+                return &(rbuf[i]);
+            }
+	}
         /* if there is a free spot: remember it */
-        if ((res == NULL) && (rbuf[i].pkt == NULL)) {
-            res = &(rbuf[i]);
-        }
+	else{
```
Comment (cgundogan) [inline sys/net/gnrc/network_layer/sixlowpan/frag/rbuf.c]: use `else if` instead

**Comment by Osblouf** [comment]
Globally yes. I was already effective I think. I also changed the second set of tests to test only one time if it is an empty spot in the loop.

**Comment by OlegHahm** [comment]
@Osblouf, thx for workin on this, but unfortunately I won't have much time to review this during the next day. So, please be patient with me...

File: sys/net/gnrc/network_layer/sixlowpan/frag/rbuf.c [hunk 2]
```diff
@@ -268,27 +268,27 @@ static rbuf_t *_rbuf_get(const void *src, size_t src_len,
...
-        /* check first if entry already available */
-        if ((rbuf[i].pkt != NULL) && (rbuf[i].pkt->size == size) &&
-            (rbuf[i].tag == tag) && (rbuf[i].src_len == src_len) &&
-            (rbuf[i].dst_len == dst_len) &&
-            (memcmp(rbuf[i].src, src, src_len) == 0) &&
-            (memcmp(rbuf[i].dst, dst, dst_len) == 0)) {
-            DEBUG("6lo rfrag: entry %p (%s, ", (void *)(&rbuf[i]),
-                  gnrc_netif_addr_to_str(l2addr_str, sizeof(l2addr_str),
+        /* check first if entry already available */ 
+	if (rbuf[i].pkt != NULL){
```
Comment (authmillenon) [inline sys/net/gnrc/network_layer/sixlowpan/frag/rbuf.c]: Whitespace missing and why did you took this out of the rest of the if-statement?

File: sys/net/gnrc/network_layer/sixlowpan/frag/rbuf.c [hunk 3]
```diff
@@ -268,27 +268,27 @@ static rbuf_t *_rbuf_get(const void *src, size_t src_len,
...
+                rbuf[i].arrival = now.seconds;
+                return &(rbuf[i]);
+            }
+	}
         /* if there is a free spot: remember it */
-        if ((res == NULL) && (rbuf[i].pkt == NULL)) {
-            res = &(rbuf[i]);
-        }
+	else if (res == NULL) {
+                res = &(rbuf[i]);
```
Comment (authmillenon) [inline sys/net/gnrc/network_layer/sixlowpan/frag/rbuf.c]: indentation

File: sys/net/gnrc/network_layer/sixlowpan/frag/rbuf.c (hunk 2 above)
Comment (cgundogan) [inline sys/net/gnrc/network_layer/sixlowpan/frag/rbuf.c]: >...and why did you took this out of the rest of the if-statement?

to not check for `(rbuf[i].pkt == NULL)` below, I assume.

Comment (authmillenon) [inline sys/net/gnrc/network_layer/sixlowpan/frag/rbuf.c]: Ah right

**Comment by OlegHahm** [comment]
Sorry for the long silence. Disappeared from my radar. Can you rebase?

**Comment by Osblouf** [comment]
Hi. Sorry, I've been really busy lately. I corrected a merge conflict. 
I am a real noob with git, feel free to tell me if I did it wrong.

**Comment by OlegHahm** [comment]
Ok, I think what you need to do is the following:
```
git fetch --all
git rebase -i origin/master
```
Assuming that you have cloned RIOT from the main repository. If not, please add it as a remote by running `git remote add riot https://github.com/RIOT-OS/RIOT` and `git fetch riot`. Then replace `origin` by `riot` in the rebase command.

Finally the rebase command should open a text editor showing you all your commits that are not yet merged into RIOT. Check if all commits are there and close the editor. It might happen now that git tells you that there are conflicts. In this case run `git mergetool`, resolve the conflicts and run `git rebase --continue`. After you're done you can run `git diff Osblouf/ng_sixlowpan` (assuming that you have added a remote called `Osblouf` pointing to your fork of RIOT) to check the diff between your local copy and this branch on Github.

If you're okay with the result you can run `git push -f Osblouf ng_sixlowpan` to force-push the local branch to Github.

**Comment by Osblouf** [comment]
Ok, I have done something like that. I have every commits made since my fork in the branch attached to this pull request.

**Comment by cgundogan** [comment]
Can you try `git rebase origin/master` again? This should hide all commits not belonging to you in this PR.

**Comment by OlegHahm** [comment]
Any luck?

**Comment by authmillenon** [comment]
Just by looking at the code I don't see how this makes things more efficient... C's `&&` is evaluated lazy anyways so your change doesn't change much.

**Comment by authmillenon** [comment]
At the last Hack'n'ACK we said we give this PR's author one week time to react and then close it. We did this so I close it.
